Make `/usr/bin/coreos-assembler` a Go program, implement `clean` in Go

cgwalters opened the pull request:
mantle: Display stderr from `ssh-keygen`

This is failing in CI, but on general principle we don't want
to swallow stderr.

---

Make `/usr/bin/coreos-assembler` a Go program, implement `clean` in Go

- Converts the entrypoint into Go code
- Add an internal library that exposes/wraps `cmdlib.sh`
  because we have a lot of stuff in there that can't be ported
  to Go yet.
- Add an internal library for running inline (named) bash scripts
- Port `clean` to Go

This is a pattern I think we'll use to aid the transition; rather
than trying to rewrite things wholesale in Go, we'll continue
to exec some shell scripts.

Gradually perhaps, we may invert some things and change both
`cmdlib.sh` and `cmdlib.py` to exec the cosa Go process in some
cases too.

Closes: https://github.com/coreos/coreos-assembler/issues/2821

---

cgwalters commented:
To restate, the high level goal here is:

- It becomes possible to use Go (alongside shell and Python) for "cosa"
- It becomes discouraged/deprecated to write new cosa code in shell or Python

There's a ton of clear next steps here; the biggest is merging mantle/ to the toplevel, so that e.g. `kola -> cosa`; i.e. `/usr/bin/cosa` is a Go program containing the kola Go code, and we dispatch on `argv[0]` etc.   We could do the same with `gangplank/`.

bgilbert commented:
> It becomes discouraged/deprecated to write new cosa code in shell or Python

I don't think there's consensus for this part.  Restructuring for convenience is one thing; setting policy is another.  Let's not block the former on the latter?

cgwalters commented:
> I don't think there's consensus for this part. Restructuring for convenience is one thing; setting policy is another. Let's not block the former on the latter?

Definitely agreed!  I'm just setting out my goals/perspective.

miabbott commented:
Would this be something that would be better suited for a dedicated branch?

Maybe it is better to keep this silo'ed from `main` until it is more fully featured?  Just asking for other opinions.

cgwalters commented:
> Would this be something that would be better suited for a dedicated branch?  Maybe it is better to keep this silo'ed from main until it is more fully featured? 

Hmm...this PR is really small I think and unlikely to break anything - CI covers almost everything.  The corner case I wrestled with turned out to be `cosa shell` where we need to carefully pass through stdin/stdout/stderr instead of buffering; which we do right now for everything but in the future won't.

Now if you're talking about the *followups* to this PR...I'm personally probably not planning on doing any wholesale or risky further followup changes in the short term.  It could make sense though to branch perhaps if anyone chose to do that kind of thing.

I think it makes the most sense to land incremental changes though as much as we can; that's really what this is.

miabbott commented:
> I think it makes the most sense to land incremental changes to `main` though as much as we can; that's really what this is.

Thanks for sharing your perspective, Colin!

travier commented:
I'm +1 for this but maybe we should have at least one command making use of it as an example?

jlebon reviewed (dismissed):
I'm also :+1: to this.

While I strongly agree we shouldn't start trying to rewrite everything in Go, I'd say at least the bash entrypoint itself would be worthwhile to rewrite to make this less confusing (and it's not large anyway).

cgwalters commented:
OK rebased :surfer: 
and this now depends on
- https://github.com/coreos/coreos-assembler/pull/2961
- https://github.com/coreos/coreos-assembler/pull/2960

cgwalters commented:
```
[2022-07-06T18:53:13.389Z] FAIL: pxe-install (bios + metal) (0s)
[2022-07-06T18:53:13.389Z]     running ssh-keygen: exit status 255
[2022-07-06T18:53:13.389Z] Error: running ssh-keygen: exit status 255
```

Huh, fascinating :detective:

cgwalters commented:
OK so it took me a while to realize that this bit of bash code was...buggy:

```
#!/usr/bin/env bash
set -euo pipefail
...
# docker/podman don't run through PAM, but we want this set for the privileged
# (non-virtualized) path
export USER="${USER:-$(id -nu)}"
```

The thing is apparently because `$(` is run in a subshell or something, bash will ignore errors there.  I think this whole time we've ended up with an empty `USER` environment variable in the OCP dynamic user case.

I adjusted the new Go logic to have a cleaner fallback here.

cgwalters commented:
:heavy_check_mark: continuous-integration/jenkins/pr-merge — This commit looks good 

:tada:

dustymabe commented:
In the past `/usr/bin/cosa` is just a convenience piece put into place that points to `/usr/bin/coreos-assembler` (via a symlink). I think we should keep that relationship rather than the other way around and also name the file `cmd/coreos-assembler.go`.

cgwalters commented:
Sure, done

dustymabe commented:
So IIUC the `bashexec` go module isn't used by anything right now? I think it would be more obvious what's going on if we converted something to use it. How about something really simple like `cmd-shell` ?

dustymabe commented on internal/pkg/bashexec/bashexec.go:
```diff
@@ -0,0 +1,60 @@
+	cmd  *exec.Cmd
+}
+
+// NewBashRunner creates a bash executor from in-memory shell script.
```
It's not immediately clear to me (I feel like I should know this but don't) why we need to convert the shell script (file) to be in memory first. What benefits does this have? Should we add a comment/explanation?

cgwalters replied:
We now have 3 different ways to interact with shell.

- `pkg/internal/cosash` is a "co-processing" proxy for `cmdlib.sh`
- `pkg/internal/bashexec` (what you're commenting on) is about a convenience wrapper for executing shell script embedded in memory in Go (which may or may not use `cmdlib.sh`, but mostly for things that aren't)
- Running fully external scripts as a subprocess, things that usually live in `/usr/lib/coreos-assembler/foo.sh`

The rationale for in-memory embedding is to aid *incremental* porting of things like `cmd-build.sh`.  Basically we might start by changing the option processing to be Go, but use Go file embedding to actually move the data from living in `/usr/lib/coreos-assembler` to being in the binary.

This makes it clearer that the goal is to use Go, not to have distinct shell scripts.

dustymabe replied:
> We now have 3 different ways to interact with shell.
> 
> * `pkg/internal/cosash` is a "co-processing" proxy for `cmdlib.sh`

I feel like I kind of understand this but it's a struggle. This is brilliant and helps us get closer to our goal but I think we're starting to violate a [basic principle](https://www.goodreads.com/quotes/273375-everyone-knows-that-debugging-is-twice-as-hard-as-writing) here:

```
“Everyone knows that debugging is twice as hard as writing a
program in the first place. So if you're as clever as you
can be when you write it, how will you ever debug it?”
```

IOW, how many people on our team fully grok what is going on here and how many people can debug it?


> * `pkg/internal/bashexec` (what you're commenting on) is about a convenience wrapper for executing shell script embedded in memory in Go (which may or may not use `cmdlib.sh`, but mostly for things that aren't)

I guess I still don't really understand the "embedded in memory" part.  Why can't we use a file on disk rather than the `fd/3` magic. I'm missing something fundamental about why that part is required. Maybe the new process doesn't have access to the old filesystem view of the world?

> * Running fully external scripts as a subprocess, things that usually live in `/usr/lib/coreos-assembler/foo.sh`

I feel like this is the easiest option to understand, I was even able to figure it out myself on Friday, but obviously we aren't able to take as "incremental" of an approach with this strategy.

> 
> 
> The rationale for in-memory embedding is to aid _incremental_ porting of things like `cmd-build.sh`. Basically we might start by changing the option processing to be Go, but use Go file embedding to actually move the data from living in `/usr/lib/coreos-assembler` to being in the binary.
> 
> This makes it clearer that the goal is to use Go, not to have distinct shell scripts.

cgwalters replied:
> I feel like I kind of understand this but it's a struggle. This is brilliant and helps us get closer to our goal 

Hah, thanks - I appreciate the complement, seriously!  I'll be honest I was quite happy when I typed out that code and it all basically worked the first time I ran it.  But, it's probably because I have a *lot* of experience with gluing together code in multiple languages and have learned the hard way what works and what doesn't.  (And really more generally, things like [what can work but is overkill and hard to debug](https://en.wikipedia.org/wiki/Common_Object_Request_Broker_Architecture))

e.g. I also have worked on [gobject introspection](https://gi.readthedocs.io/en/latest/) and https://github.com/containers/containers-image-proxy-rs 

> but I think we're starting to violate a [basic principle](https://www.goodreads.com/quotes/273375-everyone-knows-that-debugging-is-twice-as-hard-as-writing) here:

I think this is a valid point.  But:

- I personally can't think of a simpler approach that will also allow us to avoid duplicating or reimplementing the shell script.  The first path would lead to a lot of extra typing and new bugs, and the latter would easily scope-creep into a wholesale rewrite that never works (and would be a continual conflict fest until it merged)
- But on the flip side, I think with *some* rewriting in the near term future we can avoid the need for the special "cosash" co-processing path and just fork off "stateless" shell script, so this code *will* go away

dustymabe replied:
> > but I think we're starting to violate a [basic principle](https://www.goodreads.com/quotes/273375-everyone-knows-that-debugging-is-twice-as-hard-as-writing) here:
> 
> I think this is a valid point. But:
> 
> * I personally can't think of a simpler approach that will also allow us to avoid duplicating or reimplementing the shell script.  The first path would lead to a lot of extra typing and new bugs, and the latter would easily scope-creep into a wholesale rewrite that never works (and would be a continual conflict fest until it merged)

Right. I think we probably do end up doing this, but an alternative is that we go with the 'duplicating' approach with a plan and a timeline for migrating everything else. The problem here is I don't think we can are in a position to commit to that.

Sorry for the noise here.

> * But on the flip side, I think with _some_ rewriting in the near term future we can avoid the need for the special "cosash" co-processing path and just fork off "stateless" shell script, so this code _will_ go away

👍

dustymabe commented on internal/pkg/bashexec/bashexec.go:
```diff
@@ -0,0 +1,60 @@
+		return nil, err
+	}
+
+	fullargs := append([]string{"-c", ". /proc/self/fd/3", name}, args...)
```
so `fd/1` and `fd/2` are `stdout`/`stderr` and we know using `fd/3` is safe because we just opened a new file with `os.CreateTemp` above. Maybe we add a comment above the `CreateTemp` that says something like "This will be file descriptor 3 (`fd/3`)".

dustymabe commented on cmd/coreos-assembler.go:
```diff
@@ -0,0 +1,170 @@
+package main
```
Might be nice to have at least a sentence or two description up here at the top of this file.

dustymabe commented on cmd/coreos-assembler.go:
```diff
@@ -0,0 +1,170 @@
+	// Set PYTHONUNBUFFERED=1 so that we get unbuffered output. We should
+	// be able to do this on the shebang lines but env doesn't support args
+	// right now. In Fedora we should be able to use the `env -S` option.
+	os.Setenv("export PYTHONUNBUFFERED", "1")
```
```suggestion
	os.Setenv("PYTHONUNBUFFERED", "1")
```

Otherwise I'm getting:

```
Error: error parsing host environment variables: name "export PYTHONUNBUFFERED" has white spaces, poorly formatted name
```

cgwalters commented on internal/pkg/cosash/cosash.go:
```diff
@@ -0,0 +1,163 @@
+}
+
+// PrepareBuild prepares for a build
+func (sh *CosaSh) PrepareBuild() error {
```
self-review (for followup): this should return the cwd of the new build.

cgwalters commented on internal/pkg/cosash/cosash.go:
```diff
@@ -0,0 +1,163 @@
+
+// HasPrivileges checks if we can use sudo
+func (sh *CosaSh) HasPrivileges() (bool, error) {
+	 r, err := sh.processWithReply(`
+if has_privileges; then
+  echo true >&3
+else
+  echo false >&3
+fi`)
```
So personally I think this is a nifty (and yes, also hacky) demonstration of how we can practically not rewrite all the shell script at once.  

To proxy a conditional like this we need to (well, want to) do something a bit more sophisticated than just fork off a process and collect its exit status.

This code demonstrates the bi-directional IPC with a running shell process; here we're returning a boolean, but we could also e.g. have the shell return JSON which we parse, etc.

dustymabe commented on cmd/clean.go:
```diff
@@ -0,0 +1,56 @@
+package main
+
+import (
+	_ "embed"
```
```suggestion
```

Where is this used?

dustymabe commented on cmd/coreos-assembler.go:
```diff
@@ -0,0 +1,178 @@
+package main
+
+import (
+	_ "embed"
```
```suggestion
```

Where is this used?

dustymabe commented on internal/pkg/bashexec/bashexec.go:
```diff
@@ -0,0 +1,83 @@
+}
+
+// Run spawns the script, gathering stdout/stderr into a buffer that is displayed only on error.
+func (r *BashRunner) Run() error {
```
Might be marginally helpful if we disambiguate this `Run()` with the one below (line 77). The function signature's are different so the compiler knows (obviously), but could trip up someone reading the code.

cgwalters commented:
:new: 

- all comments above addressed
- both `bashexec` and `cosash` inject bash strict mode now
- Added basic unit tests for bashexec
- Added `Makefile` glue to execute go tests for the above

cgwalters commented:
Nice, of course it randomly failed with a seemingly unrelated dynamic Python traceback...
Not reproducing...oh.  Oh wait, I think this is fallout from recent changes to the CI config for openshift/os.  We're just doing `cosa build ostree` and not `cosa build`.  And *of course* that results in a dynamic Python crash when you `cosa buildextend-aliyun`.

cgwalters commented:
Hmm, timeout
/retest

cgwalters commented:
> I noticed this is marked as closing https://github.com/coreos/coreos-assembler/issues/2821 but I'm not sure we should tie the two. That cosa issue seems more to be a project policy decision.

There's a lot of fuzzy lines here...the title of that issue is just "default to Go" which one could interpret in multiple ways.

To me, "default to Go" is literally that the first code executed by `/usr/bin/coreos-assembler` is Go.

cgwalters commented:
CI is probably fixed by https://github.com/openshift/os/pull/891
/retest

cgwalters commented:
/retest

travier commented:
> @travier @mike-nguyen as TLs mind weighing in officially on this? +1 / -1 / 0 / other?

I'm late here but I already agreed in principle in https://github.com/coreos/coreos-assembler/pull/2919#issuecomment-1159015252 and in general I'm in favor of us moving code from Bash or Python to Go.
